pkg/block/indexheader/binary_reader.go

```diff
@@ -0,0 +1,802 @@
+	postings map[string][]postingOffset
+	// For the v1 format, labelname -> labelvalue -> offset.
+	postingsV1 map[string]map[string]index.Range
+
+	symbols     *index.Symbols
+	nameSymbols map[uint32]string // Cache of the label name symbol lookups,
+	// as there are not many and they are half of all lookups.
+
+	dec *index.Decoder
+
+	version             int
+	indexVersion        int
+	indexLastPostingEnd int64
+}
+
+// NewBinaryReader loads or builds new index-header if not present on disk.
+func NewBinaryReader(ctx context.Context, logger log.Logger, bkt objstore.BucketReader, dir string, id ulid.ULID) (*BinaryReader, error) {
+	binfn := filepath.Join(dir, id.String(), block.IndexHeaderFilename)
+	br, err := newFileBinaryReader(binfn)
+	if err == nil {
+		return br, nil
+	}
+
+	level.Warn(logger).Log("msg", "failed to read index-header from disk; recreating", "path", binfn, "err", err)
+
+	if err := WriteBinary(ctx, bkt, id, binfn); err != nil {
+		return nil, errors.Wrap(err, "write index header")
+	}
+
+	level.Debug(logger).Log("msg", "build index-header file", "path", binfn, "err", err)
+
+	return newFileBinaryReader(binfn)
+}
+
+func newFileBinaryReader(path string) (bw *BinaryReader, err error) {
+	f, err := fileutil.OpenMmapFile(path)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			runutil.CloseWithErrCapture(&err, f, "index header close")
+		}
+	}()
+
+	r := &BinaryReader{
+		b:        realByteSlice(f.Bytes()),
+		c:        f,
+		postings: map[string][]postingOffset{},
+	}
+
+	// Verify header.
+	if r.b.Len() < headerLen {
+		return nil, errors.Wrap(encoding.ErrInvalidSize, "index header's header")
+	}
+	if m := binary.BigEndian.Uint32(r.b.Range(0, 4)); m != MagicIndex {
+		return nil, errors.Errorf("invalid magic number %x", m)
+	}
+	r.version = int(r.b.Range(4, 5)[0])
+	r.indexVersion = int(r.b.Range(5, 6)[0])
+
+	r.indexLastPostingEnd = int64(binary.BigEndian.Uint64(r.b.Range(6, headerLen)))
+
+	if r.version != BinaryFormatV1 {
+		return nil, errors.Errorf("unknown index header file version %d", r.version)
+	}
+
+	r.toc, err = newBinaryTOCFromByteSlice(r.b)
+	if err != nil {
+		return nil, errors.Wrap(err, "read index header TOC")
+	}
+
+	r.symbols, err = index.NewSymbols(r.b, r.indexVersion, int(r.toc.Symbols))
+	if err != nil {
+		return nil, errors.Wrap(err, "read symbols")
+	}
+
+	var lastKey []string
+	if r.indexVersion == index.FormatV1 {
+		// Earlier V1 formats don't have a sorted postings offset table, so
+		// load the whole offset table into memory.
+		r.postingsV1 = map[string]map[string]index.Range{}
+
+		var prevRng index.Range
+		if err := index.ReadOffsetTable(r.b, r.toc.PostingsOffsetTable, func(key []string, off uint64, _ int) error {
+			if len(key) != 2 {
+				return errors.Errorf("unexpected key length for posting table %d", len(key))
+			}
+			// TODO(bwplotka): This is wrong, probably we have to sort.
+			if lastKey != nil {
+				prevRng.End = int64(off + 4)
```

Comment:
In a lot of places we add `4` to different offsets. I guess that's because of how big crc32 is. I think the code's readability could be improved if we would move that constant `4` into a variable. `crc32.Size` could probably be reused.

Reply:
done, but it's not crc, but size of length of posting.